batch/src/main/java/org/openmrs/analytics/JdbcFetchUtil.java

```diff
@@ -0,0 +1,205 @@
+				            
+				            @Override
+				            public void setParameters(KV<String, Integer> element, PreparedStatement preparedStatement)
+				                    throws Exception {
+					            String[] range = Objects.requireNonNull(element.getKey()).split(",");
+					            preparedStatement.setInt(1, Integer.parseInt(range[0]));
+					            preparedStatement.setInt(2, Integer.parseInt(range[1]));
+				            }
+			            }).withOutputParallelization(false)
+			            .withQuery(
+			                String.format("select uuid from %s where %s >= ? and %s < ?", this.tableName, tableId, tableId))
+			            .withRowMapper(new JdbcIO.RowMapper<String>() {
+				            
+				            @Override
+				            public String mapRow(ResultSet resultSet) throws Exception {
+					            return resultSet.getString("uuid");
+				            }
+			            }).withCoder(NullableCoder.of(StringUtf8Coder.of())));
+		}
+	}
+	
+	/**
+	 * batch together request using batchSize and generate segment descriptors `_id?<uuid>,<uuid>..`
+	 */
+	public static class CreateSearchSegments extends PTransform<PCollection<String>, PCollection<SearchSegmentDescriptor>> {
+		
+		private final String baseBundleUrl;
+		
+		private final String resourceType;
+		
+		private final int batchSize;
+		
+		public CreateSearchSegments(String resourceType, String baseBundleUrl, int batchSize) {
+			this.baseBundleUrl = baseBundleUrl;
+			this.batchSize = batchSize;
+			this.resourceType = resourceType;
+		}
+		
+		@Override
+		public PCollection<SearchSegmentDescriptor> expand(PCollection<String> uuIds) {
+			return uuIds
+			        // create KV required by GroupIntoBatches
+			        .apply(ParDo.of(new DoFn<String, KV<String, String>>() {
+				        
+				        @ProcessElement
+				        public void processElement(ProcessContext c) {
+					        if (c.element() != null) {
+						        c.output(KV.of(resourceType, c.element()));
+					        }
+				        }
+			        }))
+			        .apply(String.format("GroupIntoBatches of %s", this.batchSize),
+			            GroupIntoBatches.<String, String> ofSize(this.batchSize))
+			        .apply("Generate Segments", ParDo.of(new DoFn<KV<String, Iterable<String>>, SearchSegmentDescriptor>() {
+				        
+				        @ProcessElement
+				        public void processElement(@Element KV<String, Iterable<String>> element,
+				                OutputReceiver<SearchSegmentDescriptor> r) {
+					        List<String> uuIds = new ArrayList<String>();
+					        element.getValue().forEach(uuIds::add);
+					        r.output(SearchSegmentDescriptor.create(
+					            String.format("%s?_id=%s", baseBundleUrl, String.join(",", uuIds)), uuIds.size()));
+					        
+				        }
+			        }));
+		}
+	}
+	
+	public ResultSet FetchMaxId(String tableName) throws SQLException {
```

Comment:
@kimaina thanks for the great work....
Should this be  `fetchMaxId()`?

Reply:
Thanks @gitcliff for the catch, yes it should be! Forgot to rename initially this was a class

Reply:
> @kimaina That's an interesting finding. There are always likely to be hard limits on the size of queries like this, since most web servers have some kind of limits to prevent DOS-style attacks. At ~200 UUIDs (which are encoded as, at best, 36 bytes), we're probably hitting Tomcat 7's default `maxHttpHeaderSize` of 8 KB (back of the napkin math suggests that 8 KB can handle 228 UUIDs, assuming absolutely no other headers get sent). My impression is that instead of rejecting the request, Tomcat just cuts it off and hands it off to the server to process, hence HAPI seeing a "Bad Request". I see three options:
> 
> 1. Mess with Tomcat's configuration to up the `maxHttpHeaderSize` for the HTTP connector.
> 2. Limit the batch size to, e.g., 150-175 UUIDs.
> 3. Use the FHIR POST search end-point (i.e., post a request to `/Observation/_search` with the `_id` parameter in the body of the request). This gets around things as the default Tomcat limit for POST bodies is 2 MB.
> 
> 2 is obviously the easiest to implement.

Thanks, Ian for these suggestions! Is option 3 currently implemented? In that case, should we go  for this option in a later iteration, to improve flexibility